app/models/physical_storage.rb

```diff
@@ -39,4 +44,81 @@ def refresh_ems
 
     EmsRefresh.queue_refresh(ext_management_system)
   end
+
+  def self.create_physical_storage_queue(userid, ext_management_system, options = {})
+    task_opts = {
+      :action => "creating Cloud PhysicalStorage for user #{userid}",
+      :userid => userid
+    }
+
+    queue_opts = {
+      :class_name  => 'PhysicalStorage',
+      :method_name => 'create_physical_storage',
+      :role        => 'ems_operations',
+      :queue_name  => ext_management_system.queue_name_for_ems_operations,
+      :zone        => ext_management_system.my_zone,
+      :args        => [ext_management_system.id, options]
+    }
+
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def self.create_physical_storage(ems_id, options = {})
+    raise ArgumentError, _("ems_id cannot be nil") if ems_id.nil?
+
+    ext_management_system = ExtManagementSystem.find(ems_id)
+    raise ArgumentError, _("ext_management_system cannot be found") if ext_management_system.nil?
+
+    klass = class_by_ems(ext_management_system)
+    klass.raw_create_physical_storage(ext_management_system, options)
+  end
+
+  def self.class_by_ems(ext_management_system)
+    # TODO(lsmola) taken from Orchestration stacks, correct approach should be to have a factory on ExtManagementSystem
+    # side, that would return correct class for each provider
+    ext_management_system && ext_management_system.class::PhysicalStorage
+  end
+
+  def self.validate_create_physical_storage(ext_management_system)
+    klass = class_by_ems(ext_management_system)
+    return klass.validate_create_physical_storage(ext_management_system) if ext_management_system &&
+                                                                            klass.respond_to?(:validate_create_physical_storage)
```

Comment:
Usually we would use `supports?(:create_physical_storage)` and have the `PhysicalStorage` provider subclass which does support it set `supports :create_physical_storage`.

We should be able to get the provider subclass for PhysicalStorage (e.g. `ManageIQ::Providers::Autosde::StorageManager::PhysicalStorage`) /cc @himdel)

Reply:
Here's an example https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/cloud_network_controller.rb#L48

Reply:
I was trying to reuse the pattern from cloud_volume_controller. Is this pattern deprecated?

https://github.com/ManageIQ/manageiq-ui-classic/blob/b79c7e69160706011b902522271cca2fb2e37875/app/controllers/cloud_volume_controller.rb#L214

Also, if using the `supports` Approach, shouldn't I write `supports :create` in stead of `supports :create_physical_storage`?

Reply:
Yeah this is the availability_mixin style which is being replaced by supports_feature_mixin https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/availability_mixin.rb#L4